librosa/segment.py

```diff
@@ -38,13 +39,213 @@
+    >>> plt.subplot(1, 2, 1)
+    >>> librosa.display.specshow(R, x_axis='time', y_axis='time')
+    >>> plt.title('Binary recurrence (symmetric)')
+    >>> plt.subplot(1, 2, 2)
+    >>> librosa.display.specshow(R_aff, x_axis='time', y_axis='time',
+    ...                          cmap='magma_r')
+    >>> plt.title('Affinity recurrence')
+    >>> plt.tight_layout()
+
+    '''
+    data1 = np.atleast_2d(data1)
+    data2 = np.atleast_2d(data2)
+
+    if data1.shape[0] != data2.shape[0]:
+        raise ValueError("data1 and data2 must have the same first dimension")
+
+    data1 = np.swapaxes(data1, -1, 0)
+    n_times1 = data1.shape[0]
+    data1 = data1.reshape((n_times1, -1))
+
+    data2 = np.swapaxes(data2, -1, 0)
+    n_times2 = data2.shape[0]
+    data2 = data2.reshape((n_times2, -1))
+
+    n_feat = data1.shape[-1]
+
+    if width < 1:
+        raise ParameterError('width must be at least 1')
+
+    if mode not in ['connectivity', 'distance', 'affinity']:
+        raise ParameterError(("Invalid mode='{}'. Must be one of "
+                              "['connectivity', 'distance', "
+                              "'affinity']").format(mode))
+    if k is None:
+        if n_times1 > 2 * width + 1:
+            k = 2 * np.ceil(np.sqrt(n_feat - 2 * width + 1))
+        else:
+            k = 2
+
+    if bandwidth is not None:
+        if bandwidth <= 0:
+            raise ParameterError('Invalid bandwidth={}. '
+                                 'Must be strictly positive.'.format(bandwidth))
+
+    k = int(k)
+
+    # Build the neighbor search object
+    try:
+        knn = sklearn.neighbors.NearestNeighbors(n_neighbors=min(n_times1-1, k + 2 * width),
+                                                 metric=metric,
+                                                 algorithm='auto')
+    except ValueError:
+        knn = sklearn.neighbors.NearestNeighbors(n_neighbors=min(n_times1-1, k + 2 * width),
+                                                 metric=metric,
+                                                 algorithm='brute')
+
+    knn.fit(data1)
+
+    # Get the knn graph
+    if mode == 'affinity':
+        kng_mode = 'distance'
+    else:
+        kng_mode = mode
+
+    xsim = knn.kneighbors_graph(X=data2, mode=kng_mode).tolil()
+
+    # Remove connections within width
+    for diag in range(-width + 1, width):
+        xsim.setdiag(0, diag)
+
+    # Retain only the top-k links per point
+    for i in range(n_times2):
+        # Get the links from point i
+        links = xsim[i].nonzero()[1]
+
+        # Order them ascending
+        idx = links[np.argsort(xsim[i, links].toarray())][0]
+
+        # Everything past the kth closest gets squashed
+        xsim[i, idx[k:]] = 0
+
+    xsim = xsim.tocsr()
+    xsim.eliminate_zeros()
+
+    if mode == 'connectivity':
+        xsim = xsim.astype(np.bool)
+    elif mode == 'affinity':
+        if bandwidth is None:
+            bandwidth = np.median(xsim.max(axis=1).data)
```

Comment:
this is estimating the median similarity across the second axis; is that what we want here?  I'm a little turned around at this point.

Reply:
the bandwidth parameter is just `normalizing` the affinity scores right? Not sure what the right answer is here.

Reply:
The bandwidth estimate heuristic we use here is the median (over queries) distance to the k'th nearest neighbor of each query.  I just want to make sure the axes here are working in the correct direction.

Reply:
Yes, I'm pretty sure the second axis is over queries here.